genie-web/src/main/java/com/netflix/genie/web/data/entities/listeners/JobEntityListener.java

```diff
@@ -0,0 +1,101 @@
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.data.entities.listeners;
+
+import com.netflix.genie.common.dto.JobStatus;
+import com.netflix.genie.web.data.entities.JobEntity;
+import com.netflix.genie.web.data.observers.PersistedJobStatusObserver;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nullable;
+import javax.persistence.PostLoad;
+import javax.persistence.PostUpdate;
+
+/**
+ * Listener for Job JPA entity ({@link JobEntity}).
+ * Currently tracks persistent changes to the status of a job and notifies an observer.
+ * Could be extended to do proxy more persist changes.
+ * <p>
+ * N.B. Spring configuration.
+ * - This class does not appear in any AutoConfiguration as bean.
+ * It is referenced as {@link java.util.EventListener} by {@link JobEntity}.
+ * EntityManager creates it even if a bean of the same type exists already.
+ * - The constructor parameter {@code persistedJobStatusObserver} is marked {@code Nullable} so that an instance can be
+ * created even if no bean of that type is configured.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class JobEntityListener {
+
+    private final PersistedJobStatusObserver persistedJobStatusObserver;
+
+    /**
+     * Constructor (autowired during Persistence / EntityManager initialization).
+     *
+     * @param persistedJobStatusObserver the observer to notify
+     */
+    public JobEntityListener(
+        @Nullable final PersistedJobStatusObserver persistedJobStatusObserver
+    ) {
+        this.persistedJobStatusObserver = persistedJobStatusObserver;
+    }
+
+    /**
+     * Persistence callback invoked after a job entity has been committed/flushed into persistent storage.
+     * If the persisted status of the job is different from the last one notified, then a notification is emitted.
+     *
+     * @param jobEntity the job that was just persisted
+     */
+    @PostUpdate
+    public void jobUpdate(final JobEntity jobEntity) {
+
+        final JobStatus currentState = jobEntity.getStatus();
+        final JobStatus previouslyNotifiedState = jobEntity.getNotifiedJobStatus();
+        final String jobId = jobEntity.getUniqueId();
+        if (currentState != previouslyNotifiedState) {
+            log.info(
+                "Detected state change for job: {} from: {} to: {}",
+                jobId,
+                previouslyNotifiedState,
+                currentState
+            );
+
+            // Notify observer if not null
+            if (this.persistedJobStatusObserver != null) {
```

Comment:
If the observer is null why bother doing anything? There doesn't appear to be any way to change it from being null as the constructor is the only modifier so why not short circuit all the other logic and just return.

Reply:
Tried to keep this listener generic.
For example there may be other observers for other parts of the entity.